### PR TITLE
Syncer: Sweep Btc

### PR DESCRIPTION
**Pull request by TheCharlatan**
Part of #493 , the sweep btc functionality is implemented to eventually allow for recovery of funds stuck in the funding address.

### CODE REVIEW

File: src/syncerd/bitcoin_syncer.rs
Comment (zkao) [inline src/syncerd/bitcoin_syncer.rs:478]: no clue, 2 is the fee rate?

File: src/syncerd/bitcoin_syncer.rs
Comment (zkao) [inline src/syncerd/bitcoin_syncer.rs:504]: this error is unreachable, insert is just above, and `pubkey`, `full_sig` are just `pk`, `sig_all`

File: src/syncerd/bitcoin_syncer.rs
Comment (zkao) [inline src/syncerd/bitcoin_syncer.rs:929]: add the else clause to if let Sweep::Bitcoin


this is like above, just copied the msg, change it, please

`else { error!("Aborting sweep address task - unable to decode sweep address addendum"); }`

**Review by zkao — COMMENTED** [review]
Just minor comments, I recommend one extra review by a third party

File: src/syncerd/bitcoin_syncer.rs
Comment (TheCharlatan) [reply src/syncerd/bitcoin_syncer.rs]: Done in 2e87d843162ea948455af10626edd4ed1ad7d206

File: src/syncerd/bitcoin_syncer.rs
Comment (TheCharlatan) [reply src/syncerd/bitcoin_syncer.rs]: Oh, this was ugly. Done in 056d1843ba99762556f6cd540faa379a3f911ee8

File: src/syncerd/bitcoin_syncer.rs
Comment (TheCharlatan) [reply src/syncerd/bitcoin_syncer.rs:929]: Done in 4ba2040b82d5f8729feae94eecebfb3376f08730

File: src/syncerd/bitcoin_syncer.rs
Comment (TheCharlatan) [reply src/syncerd/bitcoin_syncer.rs:479]: Sorry, I should have left a comment there, since I wanted to discuss this. Yes, it is the fee rate. I don't know how I should handle it properly yet. Should we add a fee priority level to the SweepAddress task?

Comment (zkao) [reply src/syncerd/bitcoin_syncer.rs:479]: In swapd i also do some ugly thing for the fees

personally i advocate for mempool position as the best way to prioritize txs

And yes, it does make sense to add priority level to sweepaddress

File: src/syncerd/bitcoin_syncer.rs
Comment (Lederstrumpf) [inline src/syncerd/bitcoin_syncer.rs:488]: document, plox

File: src/syncerd/bitcoin_syncer.rs
Comment (Lederstrumpf) [reply src/syncerd/bitcoin_syncer.rs]: 416-424 (now 421-429) is same as https://github.com/rust-bitcoin/rust-bitcoin/commit/d882b68a2cbba8e643695552f38e4b36558f9617.

`sweep_address` lgtm, but good if @h4sh3d double checks.

File: src/syncerd/bitcoin_syncer.rs
Comment (TheCharlatan) [reply src/syncerd/bitcoin_syncer.rs]: Done in https://github.com/farcaster-project/farcaster-node/pull/499/commits/e0e7aa9bb7f17ee1a314900ade88589e6ded3097

Comment (TheCharlatan) [reply src/syncerd/bitcoin_syncer.rs]: Misplaced a parenthesis -_- fixed in https://github.com/farcaster-project/farcaster-node/pull/499/commits/62fe50216ca1586a83bb6720e6148579282cf389

**Review by Lederstrumpf — APPROVED** [review]
tACK @ 62fe50216ca1586a83bb6720e6148579282cf389.

File: src/syncerd/bitcoin_syncer.rs
Comment (TheCharlatan) [reply src/syncerd/bitcoin_syncer.rs:437]: Good point, it's an artefact from the Monero code where splitting into multiple transactions when there are too many outputs is  handled by the monero wallet. Thus the syncer state interface requires a `Vec<Vec<_>>` for the sweep handling. Unlike Monero where the maximum number of inputs is ~1000, Bitcoin sweep transactions (with `n` inputs and 1 output) can have ~20'000 inputs, so there is no realistic need for splitting a sweep into multiple transactions.

File: src/syncerd/bitcoin_syncer.rs
Comment (zkao) [inline src/syncerd/bitcoin_syncer.rs:456]: log nothing to sweep

**Review by zkao — APPROVED** [review]
went carefully throught the rust-bitcoin part, and it lgtm

**Review by h4sh3d — APPROVED** [review]

File: src/syncerd/bitcoin_syncer.rs
Comment (h4sh3d) [reply src/syncerd/bitcoin_syncer.rs]: lgtm

File: src/syncerd/bitcoin_syncer.rs
Comment (TheCharlatan) [reply src/syncerd/bitcoin_syncer.rs]: 426bcabd9a3ff160b1feb3991b3bcd8a5111ffd4

File: src/syncerd/bitcoin_syncer.rs
Comment (TheCharlatan) [reply src/syncerd/bitcoin_syncer.rs:420]: Will do once merged.

**Pull request merged**
